### ISSUE TITLE
How to use cnn for seq2seq tasks ? 

### ISSUE THREAD

**Issue by vivanov879** [locuslab/TCN#3]
Hi. How do I use cnn for seq2seq? I understand that cnn can be used as a encoder. What about decoder? I looked through tensorflow transformer, where you iteratively generate one symbol a time by feeding <start symbol, none, none, ...>, then <start symbol, 1st generated symbol, none, none ,...> and so on into decoder and use a mask to avoid backward information flow. Is that the same approach here ?

**Comment by jerrybai1995** [comment]
The most important thing about decoder is that you shouldn't use future information, which means it's going to be similar to the causal convolution used in TCN here. So you probably want to use a masked convolution, or you can add a padding of (k-1)*dilation (where k is the kernel size) when doing convolution, and then remove last k elements from the end of the output. 

Hope this helps :-)

**Comment by vivanov879** [comment]
Jerry, thanks for the explanation. I read that, when predicting next sequence elements, there is a local optimum that predicts exactly the previous element. As far as i know, there are 2 ways dealing with that:
1. seq2seq
2. curriculum learning: first learn to predict the next element using ground truth previous element. Then, when training, feed in ground truth element every 2 elements, all the other time use predicted element from previous step as input. Then, use ground truth only every 3rd element, then 4th, etc.
Curriculum learning seems to be less powerful than seq2seq as curriculum only uses an encoder to predict while seq2seq uses decoder as well.
I want to try TCN with curriculum learning -- do you think that will work? I'll let you know how it works.

**Comment by vivanov879** [comment]
As an unsupervised task, I also thought of picking several future values, say, [t+5, t+10, t+20] as targets at every step -- again, without a decoder.
